src/t8_schemes/t8_scheme_common_cxx.hxx

```diff
@@ -0,0 +1,631 @@
+
+class t8_scheme_common_c: public t8_eclass_scheme_c {
+ private:
+  sc_mempool_t *mempool;
+
+ public:
+  /** Destructor for all default schemes */
+  virtual ~t8_scheme_common_c ()
+  {
+    T8_ASSERT (mempool != NULL);
+    SC_ASSERT (mempool->elem_count == 0);
+    sc_mempool_destroy (mempool);
+  }
+  t8_scheme_common_c (t8_eclass_t eclass_in, int elem_size)
+  {
+    element_size = elem_size;
+    mempool = sc_mempool_new (element_size);
+    eclass = eclass_in;
+  }
+
+  /** Use a mempool to get allocate */
+  virtual void
+  t8_element_new (int length, t8_element_t **elem) const
+  {
+    T8_ASSERT (mempool != NULL);
+    T8_ASSERT (0 <= length);
+    T8_ASSERT (elem != NULL);
+    for (int i = 0; i < length; ++i) {
+      elem[i] = (t8_element_t *) sc_mempool_alloc (mempool);
+    }
+  }
+
+  virtual void
+  t8_element_destroy (int length, t8_element_t **elem) const
+  {
+    T8_ASSERT (mempool != NULL);
+    T8_ASSERT (0 <= length);
+    T8_ASSERT (elem != NULL);
+    for (int i = 0; i < length; ++i) {
+      sc_mempool_free (mempool, elem[i]);
+    }
+  }
+#if T8_ENABLE_DEBUG
+  virtual void
+  t8_element_debug_print (const t8_element_t *elem) const
+  {
+    char debug_string[BUFSIZ];
+    t8_element_to_string (elem, debug_string, BUFSIZ);
+    t8_debugf ("%s\n", debug_string);
+  }
+#endif
+
+  virtual void
+  t8_element_general_function (const t8_element_t *elem, const void *indata, void *outdata) const
+  {
+  }
+
+  virtual t8_eclass_t
+  t8_element_child_eclass (int childid) const
+  {
+    return eclass;
+  }
+
+  virtual int
+  t8_element_is_family (t8_element_t **fam) const override
+  {
+    if (t8_element_level (fam[0]) == 0)
+      return 0;
+    t8_element_t *parent, *parent_compare;
+    t8_element_new (1, &parent);
+    t8_element_new (1, &parent_compare);
+    t8_element_parent (fam[0], parent);
+    int num_children = t8_element_num_children (parent);
+    bool is_family = true;
+    for (int ichild = 1; ichild < num_children; ichild++) {
+      t8_element_parent (fam[ichild], parent_compare);
+      if (!t8_element_equal (parent, parent_compare)) {
+        is_family = false;
+        break;
+      }
+    }
+    t8_element_destroy (1, &parent);
+    t8_element_destroy (1, &parent_compare);
+    return is_family;
+  }
+
+  t8_linearidx_t
+  t8_element_linear_id_recursive (t8_element_t *elem, const t8_linearidx_t id, const int level) const
+  {
+    if (t8_element_level (elem) == 0)
+      return id;
+
+    const int childid = t8_element_child_id (elem);
+    t8_element_parent (elem, elem);
+
+    t8_linearidx_t parent_id = 0;
+    for (int ichild = 0; ichild < childid; ichild++) {
+      t8_element_child (elem, ichild, elem);
+      t8_linearidx_t num_child_descendants = t8_element_count_leaves (elem, level);
+      t8_element_parent (elem, elem);
+      parent_id += num_child_descendants;
+    }
+    parent_id += id;
+    return t8_element_linear_id_recursive (elem, parent_id, level);
+  }
+
+  void
```

Comment:
- [ ] Document this function
- [ ] This function is a helper funciton, if I understand it correctly. Can we make it static or is it used elsewhere?

Reply:
I am not sure if we can make it static, because it is a member function of the class. I have moved it into the private part for now.